### PR TITLE
torch: add expression for Torch7 the machine-learning framework

### PR DESCRIPTION
**Pull request by sergei-mironov**
This change adds support for Torch7 which is a collection of Lua packages run by luajit interpreter. Instead of using Lua infrastructure from lua-packages.nix, the hand-made Lua wrappers are used. The main reason for this - lack of luarocks support in lua-packages.nix.

To try Torch, install Torch REPL with nix-env -i trepl
There is an expression for ipython-like interface named iTorch, but it doesn't build yet, unfortunately.

### CODE REVIEW

**Comment by joachifm** [comment]
Would you please change the commit subject to `torch: init at version`.

Also, the travis failure looks like a legitimate build error.

**Comment by grwlf** [comment]
I use Torch taken from github directly. Is it OK to use 'torch: init at GITHASH' pattern in commit subject? Yes, the error looks valid, I am working on it.

**Comment by joachifm** [comment]
Use the format `pkgname-git-2016-03-13`, where the date is the timestamp of the revision you are using. See https://nixos.org/nixpkgs/manual/#sec-package-naming for details.

**Comment by grwlf** [comment]
Looks like it is fine now

File: pkgs/applications/science/machine-learning/torch/default.nix [hunk 1]
```diff
@@ -0,0 +1,12 @@
+{ pkgs, fetchgit } :
+let
+  src = fetchgit {
+    url = "https://github.com/grwlf/torch-distro";
+    # rev = "fa3a02066e5aeb08dae3bde970384e0a96887710";
+    # sha256 = "0rin42dmgyl935rvin7dq6bnav2gj84xz1idzyf9by4p3zb7w3cr";
```
Comment (globin) [inline pkgs/applications/science/machine-learning/torch/default.nix]: These seem to be left-overs

File: pkgs/applications/science/machine-learning/torch/torch-distro.nix [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+, src ? ./.
+} :
+
+let
+
+  inherit (pkgs) lib stdenv callPackage writeText readline makeWrapper
+    less ncurses cmake openblas coreutils fetchgit libuuid czmq openssl;
+
+  trace = builtins.trace;
+  # trace = with lib; const id;
```
Comment (globin) [inline pkgs/applications/science/machine-learning/torch/torch-distro.nix]: More left-overs

File: pkgs/applications/science/machine-learning/torch/torch-distro.nix [hunk 3]
```diff
@@ -0,0 +1,358 @@
+{ pkgs ? import <nixpkgs> {}
```
Comment (globin) [inline pkgs/applications/science/machine-learning/torch/torch-distro.nix]: You should probably use callPackage and specify the deps explicitly

File: pkgs/applications/science/machine-learning/torch/default.nix (hunk 1 above)
Comment (grwlf) [inline pkgs/applications/science/machine-learning/torch/default.nix]: OK, will clean them out

**Comment by globin** [comment]
Haven't checked everything thoroughly but have some small comments, see above. But great work definitely :tada:

File: pkgs/applications/science/machine-learning/torch/torch-distro.nix (hunk 3 above)
Comment (grwlf) [inline pkgs/applications/science/machine-learning/torch/torch-distro.nix]: I am trying to keep the ability of using torch-distro.nix as a standalone build script. I think, pkgs could be easily replaced with explicit deps, but `src` argument is still important for me.

**Comment by grwlf** [comment]
Fix issues pointed out by @globin

**Comment by grwlf** [comment]
Here them are! The hotfixes :)

**Comment by joachifm** [comment]
It's quite unfortunate that the lua framework cannot be re-used, but I suppose that's something that can be improved later. If this is useful in its current state, I think we should just merge it. Let me know when you're ready.

**Comment by grwlf** [comment]
Thanks. I decided to try to integrate Torch with lua-packages to some extent. Luarocks can't be moved easily, but I think I will be able to merge the luajit compiler.

**Comment by grwlf** [comment]
By the way, I checked the current version of Torch expression against [char-rnn](http://karpathy.github.io/2015/05/21/rnn-effectiveness/ ) project, and it works!

**Comment by grwlf** [comment]
Travis failed to run the build due to some unrelated file `.../chromium/source/update.nix` which is missing. Could you please re-check?

Besides that, I think the request is ready for merging. It now shares luajit with luajitPackages which is an instance of luaPackages. This way, we may combine packages installed into ~/.luarocks with torch libaraies.

**Comment by joachifm** [comment]
Travis restarted. I'll merge this once it completes, seeing as there's been no objections to this change thus far, though I hope that the code is refactored at some point (working code that exists beats perfect code that doesn't, of course).

**Comment by ilanacarra** [comment]
Travis is ill. This time it decided to build everything starting from coreutils. The job exceeded the maximum time limit for jobs, and has been terminated.

**Comment by grwlf** [comment]
Travis is ill. This time it decided to rebuild everything including coreutils.The job exceeded the maximum time limit for jobs, and has been terminated

**Comment by joachifm** [comment]
One final suggestion: add the main applications to the top-level package set, to improve discoverability. As far as I can tell, trepl is the only one that works; what about exposing it as `torch-repl` (or even use the original attrname, whatever works)?

**Comment by grwlf** [comment]
Good idea. I added new commit.

**Comment by grwlf** [comment]
Unfortunately, introducing luajitPackages was a bit optimistic. Travis shows that some packages doesn't work with luajit. For example, lua-zlib links with -llua which AFAIK is not required when using luajit. I'll revert this change.

**Comment by grwlf** [comment]
Looks like Travis reports false-negative again. The job exceeded the maximum time limit for jobs, and has been terminated. Could you please recheck it?

**Comment by joachifm** [comment]
Yeah, it wants to rebuild the kernel and so will time out, restarting it won't help until hydra catches up, I think

**Comment by grwlf** [comment]
Is there something I can do to improve the situation with Travis?

**Comment by joachifm** [comment]
Restarted travis

**Comment by grwlf** [comment]
Sadly, same thing.. rebuilding the kernel

**Comment by joachifm** [comment]
I think is mainly an issue with the branches seeing lots of heavy rebuilds lately. There's no way torch would trigger a kernel build, or am I blind?

**Comment by joachifm** [comment]
I think we can go ahead with this, if you can verify that 1) nox-review passes on master; 2) that you've performed a properly sandboxed build; 3) and tested that the output of the top-level attr works.

**Comment by joachifm** [comment]
I merged & built this locally; it works, but I get these warnings during the build of torch-repl:
```
sh: /nix/store/acqx4m997cvnc9d5ylgfiryvz3p2kiwb-trepl/bin/lua: No such file or directory
No existing manifest. Attempting to rebuild...
Missing dependency for trepl scm-1: penlight scm-1
Missing dependency for trepl scm-1: torch scm-1
Missing dependency for trepl scm-1: penlight scm-1
Missing dependency for trepl scm-1: torch scm-1
Updating manifest for /nix/store/acqx4m997cvnc9d5ylgfiryvz3p2kiwb-trepl/lib/luarocks/rocks
```
Is this just noise or do they indicate a problem with the build recipe? If not, I'm happy to merge this, seeing that it at least builds & runs for me.

**Comment by grwlf** [comment]
Regarding warnings, I think it is the consequences of the chosen build strategy: from the point of view of luarocks, we create new 'tree' for each package. 'Tree' should contain manifest describing whats in, which should be re-created.

After all, `penlight` library is available from the th, for example:

```
th> require "pl.url"
{
  unquote : function: 0x410a77b8
  quote : function: 0x410a78d0
}
```

**Comment by grwlf** [comment]
Regarding nix-review and sandboxed builds - I didn't try these tools yet, but I think I should find time to learn them. Currently, I switch the system to the version of nixpkgs, containing the torch changes and rebuild it, then install torch-repl to the local user's profile.

Besides that, I have a fork of [torch-distro](https://github.com/grwlf/torch-distro) repo. The fork contains a standalone expression `build3.nix`, which uses <nixpkgs/torch-distro.nix> with src set to `./.` . This approach makes it possible to debug the expression easily.

**Comment by joachifm** [comment]
Okay. Sandboxed builds tend to expose lots of problems, such as implicit dependencies, and so lowers the bar for merging changes. Anyway, this worked for me, so I'm just going to merge it.

**Comment by grwlf** [comment]
Thanks! Let me know if something stop working.

**Pull request merged**